Add Nextflow language support

pditommaso opened the pull request:
This PR adds the support for [Nextflow](https://www.nextflow.io/) programming language that is a DSL for computational workflows.

Current public usage is over 1k hits, see [here](https://github.com/search?utf8=%E2%9C%93&q=extension%3Anf+NOT+nothack&type=Code), [here](https://github.com/search?utf8=%E2%9C%93&q=extension%3Anf+input&type=Code) and [here](https://github.com/search?utf8=%E2%9C%93&q=extension%3Anf+output&type=Code).

Nextflow is actively developed and it has been published in [Nature Biotechnology](https://rdcu.be/qZVo).

pditommaso commented:
Hi, Is anything it can be done to improve this PR. Any feedback is welcome.

pchaigno commented:
Thanks for the pull request!

A few things:
- Where are the sample files from? (We try to keep track of their origin and licenses, and permissive licenses are preferred.)
- We also prefer sample files from "real" applications since they will be used to train the Bayesian classifier. The first one looks ok, but the second looks like it was written specifically for this pull request :-) Maybe we should replace the second one with one found on github.com...?
- Just to confirm, there isn't a TextMate, Sublime Text, or Atom grammar for Nextflow that you know of?

Looks good overall :+1:

pditommaso commented:
Thanks for your quick reply! 

> Where are the sample files from?

Do you mean the ones in this PR? I'm the author. I can add an authorship a licence header at the beginning if you want. 

> We also prefer sample files from "real" applications since they will be used to train the Bayesian classifier. 

Cool. Yes, of course you can use any of the script returned by the linked search.

> Just to confirm, there isn't a TextMate, Sublime Text, or Atom grammar

Unfortunately no. I would like to create at some point.

pditommaso commented:
Hi, I've added more example, licence and authorship info. I've added also an Atom grammar. Thanks!

pchaigno commented on lib/linguist/languages.yml:
```diff
@@ -2882,6 +2882,18 @@ NewLisp:
+Nextflow:
+  type: programming
+  ace_mode: text
+  tm_scope: none
```
You'll need to change this to `source.nextflow` to actually use the grammar.

pchaigno commented on lib/linguist/languages.yml:
```diff
@@ -2882,6 +2882,18 @@ NewLisp:
   language_id: 247
+Nextflow:
+  type: programming
+  ace_mode: text
```
Could we use `groovy` as a substitute Ace grammar or is it too far off?

pditommaso replied:
What exactly is the impact of this? what's the `ace_mode` ?

pchaigno replied:
Ooh, yeah, forgot to give context :/
It's used to highlight code in the GitHub.com [editor](https://ace.c9.io/) (when users edit files or Gists directly in the web interface).

pditommaso commented:
Regarding the license, would it be better if I use a difference license template?

pditommaso commented:
OK. I've fixed the conflicts, changed `ace_mode` and `tm_scope` as suggested. Finally I've updated the MIT license text to make more clear the authorship assignments.

pditommaso commented:
Sure thing! done!

pchaigno commented:
Thanks @pditommaso! It looks good :+1: 

@lildude Any idea why Travis-CI wants us to whitelist the grammar? Its license is correctly recognized as MIT by Licensee...

lildude commented:
> @lildude Any idea why Travis-CI wants us to whitelist the grammar? Its license is correctly recognized as MIT by Licensee...

Yup, it's because the ref for the submodule predates the recent changes to the LICENSE.md that made it clearer this is licensed under MIT:

Current ref for the grammar (19beda3): 

```
irb(main):009:0> project = Licensee::FSProject.new("./vendor/grammars/atom-language-nextflow/", detect_readme: true)
=> #<Licensee::FSProject:0x007fc1181f3ed0 @pattern="*", @dir="./vendor/grammars/atom-language-nextflow/", @detect_packages=false, @detect_readme=true>
irb(main):010:0> project.license
=> nil
irb(main):011:0>
```

Current master for the grammar: 

```
irb(main):011:0> project = Licensee::FSProject.new("./vendor/grammars/atom-language-nextflow/", detect_readme: true)
=> #<Licensee::FSProject:0x007fc11726f388 @pattern="*", @dir="./vendor/grammars/atom-language-nextflow/", @detect_packages=false, @detect_readme=true>
irb(main):012:0> project.license
=> #<Licensee::License key=mit>
irb(main):013:0>
```

@pditommaso if you update the submodule reference for the grammar, you should find the tests will pass.

Sidenote: I am working on updating Linguist to use recent Licensed and Licensee gems, however it requires a few more changes than a simple bump of the versions so it taking a bit longer than normal.

pditommaso commented:
Done!

pchaigno commented on lib/linguist/languages.yml:
```diff
@@ -2905,6 +2905,19 @@ NewLisp:
+  extensions:
+  - ".nf"
+  filenames:
+  - "main.nf"
```
This is already included in `.nf`. Any reason to add it?

pditommaso replied:
if `filenames` doesn't have any special semantic no.

pchaigno replied:
`filenames` is just a list of... well, filenames to recognize as the given language. For example, it includes `.bashrc` for Shell. I think you can remove it.

pditommaso replied:
sure thing. done.

pditommaso commented:
Hi there, it should be OK on my side. Let me know if you need further changes to get this approved.

lildude commented on lib/linguist/languages.yml:
```diff
@@ -2905,6 +2905,18 @@ NewLisp:
+  extensions:
+  - ".nf"
+  filenames:
+  - "nextflow.config"  
```
Can you please add a good illustrative sample of a `nextflow.config` file to `samples/Nextflow/filenames/` too.

Another question: would one normally want this file to count towards the language statistics? If not, you may want to mark it as vendored too.

pditommaso replied:
Also, since `nextflow.config` has not the same syntax as nextflow script I guess it should not be included in the `filenames`, should it?

lildude replied:
It really depends if it should be counted in the Nextflow language stats or not. If not, then no, we don't need the entry here and its language will be determined as per normal, unless marked as vendored or generated.

pditommaso replied:
It think it's fine to include in the language stats.

lildude replied:
:cool: Now would people expect it to be classified as "Nextflow" or the actual language it's written in? If the latter, what is that language? It appears to be a very similar to JSON from the few I've seen.

pditommaso replied:
It's a customisation of the Groovy ConfigObject syntax (hence I would propose to classify it as Nextflow)

lildude replied:
:+1: In that case, we just need an illustrative sample of a `nextflow.config` file added to `samples/Nextflow/filenames/` and we should be good to go.

lildude replied:
And by "illustrative sample", I mean a good real-world-usage file that demonstrates the usage of the file well.

pditommaso replied:
Sure thing. The one just uploaded should be illustrative enough.

pditommaso commented:
Great! Thanks to you for the great collaboration! 👏 👏 👏 

One question, how I will know when these changes will be actually applied?

lildude commented:
> One question, how I will know when these changes will be actually applied?

There's now easy way within GitHub other than to make a change to the repo and see if your language is now recognised, or check the version string at the bottom of https://github-lightshow.herokuapp.com/.

I plan to make a new release soon, however we've got a few more PRs to merge before then.

pditommaso commented:
Sure. Thanks for you feedback.

pchaigno commented:
@pditommaso Thank you for your contribution!

pditommaso commented:
🎉 🎉 🎉

pditommaso commented:
Quick question, I've just pushed come improvements to the Nextlow grammar. Will it be automatically recognised/pulled by GitHub or do I need to open a pull request ?

lildude commented:
Linguist will pick it up automatically the next time we do a full grammars update. I'll probably do one in the next week as we've got another release on the cards to address a few hiccoughs.

pditommaso commented:
Perfect. Thanks a lot.

pditommaso commented:
I thinking linguist 6.0.1 is not using the latest NF grammar available. Is there a way to check what commit ID is currently used ?

lildude commented:
The release PR https://github.com/github/linguist/pull/4016 updated the grammar to https://github.com/nextflow-io/atom-language-nextflow/tree/557669e2aea4a4ea3ce789e97435db6e91ca3eca and should have been included in the gem.

pditommaso commented:
OK, that looks fine. However the syntax highlighting seems not reported correctly by the GitHub.  For example [here](https://github.com/nextflow-io/rnaseq-nf/blob/master/main.nf#L76-L90).

The [nextflow.cson](https://github.com/nextflow-io/atom-language-nextflow/blob/master/grammars/nextflow.cson) imports the [groovy.cson](https://github.com/nextflow-io/atom-language-nextflow/blob/master/grammars/groovy.cson) file. I'm wondering if GitHub syntax highlighter is able to include import from one grammar file to another?

pditommaso commented:
Sorry if I'm insisting on this, however nextflow syntax is not properly rendered by GitHub. I'm suspecting that the problem could be that the [nextflow.cson](https://github.com/nextflow-io/atom-language-nextflow/blob/master/grammars/nextflow.cson) includes the [groovy.cson](https://github.com/nextflow-io/atom-language-nextflow/blob/master/grammars/groovy.cson) as a separate file. 

Do you think this could be the problem?

lildude commented:
> Sorry if I'm insisting on this, however nextflow syntax is not properly rendered by GitHub. I'm suspecting that the problem could be that the nextflow.cson includes the groovy.cson as a separate syntax file.
> 
> Do you think this could be the problem?

I don't think so. Quite a few other grammars pull in separate syntax files and we've not received any reports of this not working to date.

@vmg am I right, or do we have a general problem with sourced syntax files not actually being pulled in since switching to the new grammar compiler?

pditommaso commented:
Any update on this? maybe would make sense to open a separate issue ?

pditommaso commented:
FYI I've opened an issue for this problem in the last comments https://github.com/github/linguist/issues/4080.
